sdk/identity/identity/src/credentials/managedIdentityCredential/utils.ts

```diff
@@ -30,3 +30,35 @@ export function mapScopesToResource(scopes: string | string[]): string | undefin
 
   return scope.substr(0, scope.lastIndexOf(DefaultScopeSuffix));
 }
+
+/**
+ * Internal type roughly matching the raw responses of the authentication endpoints.
+ *
+ * @internal
+ */
+export interface TokenResponseParsedBody {
+  token?: string;
+  access_token?: string;
+  refresh_token?: string;
+  expires_in: number;
+  expires_on?: number | string;
+}
+
+/**
+ * Given a token response, return the expiration timestamp as the number of milliseconds from the Unix epoch.
+ * @param body A parsed response body from the authentication endpoint.
+ */
+export function parseExpiresOn(body: TokenResponseParsedBody): number {
+  if (typeof body.expires_on === "number") {
+    return body.expires_on * 1000;
+  } else if (typeof body.expires_on === "string") {
+    const asNumber = +body.expires_on;
+    if (!isNaN(asNumber)) {
+      return asNumber * 1000;
+    } else {
+      return Date.parse(body.expires_on);
```

Comment:
perhaps we should check isNaN here again and throw?

Reply:
Yes that would be ideal if the date isn't valid.

Reply:
It should have a similar format to checking for retry-after like the following:

```typescript
function parseRetryAfter(headerValue: string | undefined): number | undefined {
  if (!isDefined(headerValue)) {
    return undefined;
  }

  // Retry-After is defined in seconds
  const number = new Number(headerValue);
  if (Number.isFinite(number) && number >= 0) {
    return number.valueOf() * 1000;
  }

  // Retry-After is defined as a Date
  const retryDate = Date.parse(headerValue);
  if (Number.isNaN(retryDate)) {
    return undefined;
  }

  const diff = retryDate - Date.now();
  return diff <= 0 ? undefined : diff;
}
```

Reply:
alright I tweaked it a bit, so we'll always do our best and throw only if everything fails (in theory `expires_in` should exist and save us, but I don't trust that type to be 100% accurate)